src/main/java/org/apache/sysml/runtime/codegen/LibSpoofPrimitives.java

```diff
@@ -560,7 +560,99 @@ public static void vectPlusAdd(double bval, double[] a, double[] c, int[] aix, i
+	 * @param c resultant vector
+	 * @param ci index of c
+	 * @param len number of processed elements
+	 * @return resultant value
+	 */
+	public static void vectXorAdd(double[] a, double bval, double[] c, int ai, int ci, int len) {
+		for( int j = ai; j < ai+len; j++, ci++)
+			c[ci] +=  ( (a[j] != 0) != (bval != 0) ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double bval, double[] a, double[] c, int ai, int ci, int len) {
+		for( int j = ai; j < ai+len; j++, ci++)
+			c[ci] += ( (bval != 0) != (a[j] != 0) ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double[] a, double bval, double[] c, int[] aix, int ai, int ci, int alen, int len) {
+		for( int j = ai; j < ai+alen; j++ )
+			c[ci + aix[j]] += ( (a[j] != 0) != (bval != 0) ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double bval, double[] a, double[] c, int[] aix, int ai, int ci, int alen, int len) {
+		for( int j = ai; j < ai+alen; j++ )
+			c[ci + aix[j]] += ( (bval != 0) != (a[j] != 0) ) ? 1 : 0;
+	}
+
+	//1. scalar vs. dense vector
+	public static double[] vectXorWrite(double[] a, double bval, int ai, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( ( a[ai+j] != 0) != (bval != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	//2. dense vector vs. scalar
+	public static double[] vectXorWrite(double bval, double[] a, int ai, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( (bval != 0) != (a[ai + j] != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	//3. dense vector vs. dense vector
+	public static double[] vectXorWrite(double[] a, double[] b, int ai, int bi, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( (a[ai + j] != 0) != (b[bi + j] != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	//4. sparse vector vs scalar
+	public static double[] vectXorWrite(double[] a, double bval, int[] aix, int ai, int alen, int len) {
+		double init = (bval != 0) ? 1 : 0;
+		double[] c = allocVector(len, true, init);
+		for( int j = ai; j < ai+alen; j++ )
+			c[aix[j]] = ( (a[j] != 0) != (bval != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	//5. scalar vs. sparse vector
+	public static double[] vectXorWrite(double bval, double[] a, int[] aix, int ai, int alen, int len) {
+		double init = (bval != 0) ? 1 : 0;
+		double[] c = allocVector(len, true, init);
+		for( int j = ai; j < ai+alen; j++ )
+			c[aix[j]] = ( (bval != 0) != (a[j] != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	//6. sparse vector vs. dense vector
+	public static double[] vectXorWrite(double[] a, double[] b, int[] aix, int ai, int bi, int alen, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++ )
+			c[j] = (b[j] != 0) ? 1 : 0; //c[] = xor(0,b[]!=0)
+	//	c[] = [1, 1, 0, 1, 0]
+	//	for( int j = ai; j < ai+alen; j++ ) //overwrite, for a[]'s nnz values
+	//		c[aix[j]] = ( (a[j] != 0) != (b[bi+aix[j]] != 0) ) ? 1 : 0;
+		for( int j = 0; j < aix.length; j++ )
+			c[aix[j]] = ( b[aix[j]] != 0) ? 0 : 1;
+
```

Comment:
The two pass, approach seems even easier for `xor` case. But, one case is failing. 
`0.0 vs actual 1.0 at 0 0`

cc @mboehm7